### PR TITLE
Feat/test calc njklm values

### LINKED ISSUE #28: Add unit test: matrix_a_star: `calculate_njklm_values`
The function `calculate_njklm_values` does not have a unit test. It should have one.

### PR DESCRIPTION
**Pull request by alani-odunlami**
## Description
A test has been added for calculate_njklm_values in test_matrix_a_star.py. 

This is a feature change and does not break any existing functionality
 
## Type of change

*You can delete options that are not relevant.*

- [ ] Bug fix - *non-breaking change*
- [x] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [x] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 1]
```diff
@@ -50,9 +51,39 @@ def test_calculate_b():
...
+    )
+
+    expected_output = spark.createDataFrame(
+        [
+            (1.0, 5.0, 4.0, 0.0),
+        ],
+        ["col1", "col2", "col3", "col4"],
+    )
+
+    expected_output = expected_output.toPandas()
```
Comment (mary-cleaton) [inline tests/matrix_a_star/test_matrix_a_star.py]: The test will be much more efficient if this was defined in Pandas rather than defined in PySpark and converted to Pandas.

To hard-code a dataset in Pandas, do this:

```python
expected_output = pd.DataFrame(
  [
    list_of_row_1_values,
    list_of_row_2_values,
    etc.
  ],
  list_of_column_names
]

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 2]
```diff
@@ -50,9 +51,39 @@ def test_calculate_b():
...
+
+    # Act
+    actual_output = ma.calculate_njklm_values(test_input)
+    test_output = actual_output
```
Comment (mary-cleaton) [inline tests/matrix_a_star/test_matrix_a_star.py]: There's no need to rename this. Just call it `test_output` from the start.

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 3]
```diff
@@ -50,9 +51,39 @@ def test_calculate_b():
...
+
+    # Assert
+    pd.testing.assert_frame_equal(test_output, expected_output)
+    # assert test_output ==  expected_output#
```
Comment (mary-cleaton) [inline tests/matrix_a_star/test_matrix_a_star.py]: Please remove commented-out code.

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 4]
```diff
@@ -94,11 +125,27 @@ def test_get_matrix_a_star():
...
+
+    """
+
+    # Arrange
+
+    # Act
+
+    # Assert
+
+
```
Comment (mary-cleaton) [inline tests/matrix_a_star/test_matrix_a_star.py]: Why is this here? It shouldn't be in this branch.

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 5]
```diff
@@ -94,11 +125,27 @@ def test_get_matrix_a_star():
...
 def test_label_x_star():
     """
     Tests that label_x_star() gives the correct output when provided with
     appropriate inputs.
     """
+
```
Comment (mary-cleaton) [inline tests/matrix_a_star/test_matrix_a_star.py]: This has been altered but shouldn't have been.

Comment (alani-odunlami) [reply tests/matrix_a_star/test_matrix_a_star.py]: Oh yes. I need to create another branch for the  Tests that label_x_star

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 6]
```diff
@@ -50,9 +51,39 @@ def test_calculate_b():
...
+    # Arrange
+    test_input = spark.createDataFrame(
+        [
+            (1.0, 2.0, 1.0, 0.0),
+            (0.0, 1.0, 1.0, 0.0),
+            (0.0, 1.0, 1.0, 0.0),
+            (0.0, 1.0, 1.0, 0.0),
+        ],
+        ["col1", "col2", "col3", "col4"],
+    )
```
Comment (mary-cleaton) [inline tests/matrix_a_star/test_matrix_a_star.py]: Any test input should reflect a real (though small and artificial) test case. This doesn't really do this for two reasons:

1. The column names are uninformative and don't reflect what real column names would look like in the input.
2. This function takes a table of delta comparisons as it's input. It's impossible for a delta comparison table to contain 2.0 as it's an indicator matrix (of an indicator matrix). So, the `test_input` here is actually impossible.

Please take a look at the output of `compare_deltas` function in `.../src/indicator_matrix/indicator_matrix.py` to see what the input of this function should look like.

**Review by mary-cleaton — CHANGES_REQUESTED** [review]
Please see code-specific comments and address these before resubmitting for review. Thanks!

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 2 above)
Comment (alani-odunlami) [reply tests/matrix_a_star/test_matrix_a_star.py]: Issue addressed.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 3 above)
Comment (alani-odunlami) [reply tests/matrix_a_star/test_matrix_a_star.py]: Removed all instances of commented out code.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 4 above)
Comment (alani-odunlami) [reply tests/matrix_a_star/test_matrix_a_star.py]: New branch created for this test.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 1 above)
Comment (alani-odunlami) [reply tests/matrix_a_star/test_matrix_a_star.py]: I saw in the docstring/comments that the function is taking a Spark DataFrame. Then processed it and then return Pandas df?

Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: Yes, but you don't have to make the test inputs and outputs the way that the function makes them. :)

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 5 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: Please make sure that on this branch this test is just a test shell.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 6 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: This is much better!

If this is exactly the same dataset as that used as the test output for `test_compare_deltas`, it could be defined once as a fixture in `conftest.py` and re-used.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 4 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: I think you removed this test entirely when tidying this up. It should remain as a test shell in this branch.

**Comment by mary-cleaton** [comment]
Branch switched over to me as developer, as @ayomide2021 was only on this project for 6 weeks.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 1 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: Converted `expected_output` to Pandas dataframe.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 4 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: Restored test shell for `get_scaled_labelled_x_star`.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 5 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: This comment was in error - `test_get_scaled_labelled_x_star` should have been a test shell (and now is), `test_label_x_star` should be a full test (and was and still is).

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 6 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: Output of `test_compare_deltas` moved to be a fixture in the whole-repo `conftest.py`. Then, called by both `test_compare_deltas` and `test_calculate_njklm_values`.

**Comment by mary-cleaton** [comment]
All comments addressed. Read for re-review.

File: conftest.py [hunk 7]
```diff
@@ -36,3 +41,51 @@ def spark_mock():
     spark_mock.mode.return_value = spark_mock
     spark_mock.save.return_value = None
     return spark_mock
+
+
+@pytest.fixture(scope="module")
+def compare_deltas_output(spark):
```
Comment (danasebo) [inline conftest.py]: please add type hints

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 8]
```diff
@@ -94,11 +136,17 @@ def test_get_matrix_a_star():
     pass
 
 
+@pytest.mark.skip(reason="test shell")
+def test_get_scaled_labelled_x_star():
```
Comment (danasebo) [inline tests/matrix_a_star/test_matrix_a_star.py]: please add type hints

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 9]
```diff
@@ -50,9 +51,50 @@ def test_calculate_b():
     assert test_output == expected_output
 
 
-@pytest.mark.skip(reason="test shell")
-def test_calculate_njklm_values():
-    pass
+def test_calculate_njklm_values(spark, compare_deltas_output):
```
Comment (danasebo) [inline tests/matrix_a_star/test_matrix_a_star.py]: please add type hints

**Review by danasebo — CHANGES_REQUESTED** [review]
please add type hints to all new functions. otherwise it looks good to me :)

File: conftest.py (hunk 7 above)
Comment (mary-cleaton) [reply conftest.py]: Done.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 8 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: Done.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 9 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py]: Done.

**Comment by mary-cleaton** [comment]
All comments addressed. However, I think it's likely best if this waits for #73 to merge into `develop`, rebases to `develop` and then returns for QA. Else, I fear some type hints might get missed.

**Comment by mary-cleaton** [comment]
Rebased to `develop`, now that PR #73 is merged in. Also corrected all the missing type hints I could see.

**Review by danasebo — APPROVED** [review]
The updated changes have been reviewed, the unit tests tested, and the files checked for the missing type hints - all looks good to me now, thanks!
